subsys/debug/cpu_load/cpu_load.c

```diff
@@ -0,0 +1,289 @@
+}
+
+static nrfx_err_t ppi_free(u8_t ch)
+{
+#ifdef DPPI_PRESENT
+	if (!IS_ENABLED(CONFIG_CPU_LOAD_USE_SHARED_DPPI_CHANNELS)
+		|| ((BIT(ch) & shared_ch_mask) == 0)) {
+		return nrfx_dppi_channel_free(ch);
+	} else {
+		return NRFX_SUCCESS;
+	}
+#else
+	return nrfx_ppi_channel_free((nrf_ppi_channel_t)ch);
+#endif
+}
+
+static void ppi_cleanup(u8_t ch_tick, u8_t ch_sleep)
+{
+	nrfx_err_t err = NRFX_SUCCESS;
+
+	if (IS_ENABLED(CONFIG_CPU_LOAD_ALIGNED_CLOCKS)) {
+		err = ppi_free(ch_tick);
+	}
+
+	if ((err == NRFX_SUCCESS) && (ch_sleep != CH_INVALID)) {
+		err = ppi_free(ch_sleep);
+	}
+
+	if (err != NRFX_SUCCESS) {
+		LOG_ERR("PPI channel freeing failed (err:%d)", err);
+	}
+}
+
+static void cpu_load_log_fn(struct k_work *item)
+{
+	u32_t load = cpu_load_get();
+	u32_t percent = load / 1000;
+	u32_t fraction = load % 1000;
+
+	cpu_load_reset();
+	LOG_INF("Load:%d,%03d%%", percent, fraction);
+	k_delayed_work_submit(&cpu_load_log, CPU_LOAD_LOG_INTERVAL);
+}
+
+static int cpu_load_log_init(void)
+{
+	k_delayed_work_init(&cpu_load_log, cpu_load_log_fn);
+	return k_delayed_work_submit(&cpu_load_log, CPU_LOAD_LOG_INTERVAL);
+}
+
+static void timer_handler(nrf_timer_event_t event_type, void *context)
+{
+	/*empty*/
+}
+
+
+int cpu_load_init(void)
+{
+	u8_t ch_sleep;
+	u8_t ch_wakeup;
+	u8_t ch_tick = 0;
+	nrfx_err_t err;
+	nrfx_timer_config_t config = NRFX_TIMER_DEFAULT_CONFIG;
+	int ret = 0;
+
+	if (ready) {
+		return 0;
+	}
+
+	config.frequency = NRF_TIMER_FREQ_1MHz;
+	config.bit_width = NRF_TIMER_BIT_WIDTH_32;
+
+	if (IS_ENABLED(CONFIG_CPU_LOAD_ALIGNED_CLOCKS)) {
+		/* It's assumed that RTC1 is driving system clock. */
+		config.mode = NRF_TIMER_MODE_COUNTER;
+		err = ppi_alloc(&ch_tick,
+		       nrf_rtc_event_address_get(NRF_RTC1, NRF_RTC_EVENT_TICK));
+		if (err != NRFX_SUCCESS) {
+			return -ENODEV;
+		}
+		nrfx_gppi_channel_endpoints_setup(ch_tick,
+		     nrf_rtc_event_address_get(NRF_RTC1, NRF_RTC_EVENT_TICK),
+		     nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_COUNT));
+		nrf_rtc_event_enable(NRF_RTC1, NRF_RTC_INT_TICK_MASK);
+	}
+
+
+	err = ppi_alloc(&ch_sleep,
+		       nrf_power_event_address_get(NRF_POWER,
+						   NRF_POWER_EVENT_SLEEPENTER));
+	if (err != NRFX_SUCCESS) {
+		ppi_cleanup(ch_tick, CH_INVALID);
+		return -ENODEV;
+	}
+
+	err = ppi_alloc(&ch_wakeup,
+		       nrf_power_event_address_get(NRF_POWER,
+						   NRF_POWER_EVENT_SLEEPEXIT));
+	if (err != NRFX_SUCCESS) {
+		ppi_cleanup(ch_tick, ch_sleep);
+		return -ENODEV;
+	}
+
+	nrfx_gppi_channel_endpoints_setup(ch_sleep,
+		  nrf_power_event_address_get(NRF_POWER,
+					      NRF_POWER_EVENT_SLEEPENTER),
+		  nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_START));
+	nrfx_gppi_channel_endpoints_setup(ch_wakeup,
+		  nrf_power_event_address_get(NRF_POWER,
+					      NRF_POWER_EVENT_SLEEPEXIT),
+		  nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_STOP));
+
+
+	err = nrfx_timer_init(&timer, &config, timer_handler);
+	if (err != NRFX_SUCCESS) {
+		return -EBUSY;
```

Comment:
`ppi_cleanup()` missing here.

Reply:
moved timer init before ppi alloc so that clean up is not needed.

Reply:
This way `CONFIG_CPU_LOAD_ALIGNED_CLOCKS` won't work.

Reply:
done, moved init after handling `CONFIG_CPU_LOAD_ALIGNED_CLOCKS ` added there ppi cleanup

Reply:
Okay. If you want to keep it this way, you need to also call `nrfx_timer_uninit()` when the allocation of `ch_sleep` or `ch_wakeup` fails.

Reply:
right, in that case it was better to extend ppi_cleanup to be able to cleanup wakeup channel, too.